nuxeo-runtime/nuxeo-runtime-migration/src/main/java/org/nuxeo/runtime/migration/MigrationServiceImpl.java

```diff
@@ -0,0 +1,286 @@
+        public boolean isSupportingMerge() {
+            return true;
+        }
+
+        @Override
+        public MigrationDescriptor clone(MigrationDescriptor orig) {
+            return new MigrationDescriptor(orig);
+        }
+
+        @Override
+        public void merge(MigrationDescriptor src, MigrationDescriptor dst) {
+            dst.merge(src);
+        }
+    }
+
+    @Override
+    public void registerContribution(Object contribution, String extensionPoint, ComponentInstance contributor) {
+        switch (extensionPoint) {
+        case CONFIG_XP:
+            registerMigrationDescriptor((MigrationDescriptor) contribution);
+            break;
+        default:
+            throw new RuntimeException("Unknown extension point: " + extensionPoint);
+        }
+    }
+
+    @Override
+    public void unregisterContribution(Object contribution, String extensionPoint, ComponentInstance contributor) {
+        switch (extensionPoint) {
+        case CONFIG_XP:
+            unregisterMigrationDescriptor((MigrationDescriptor) contribution);
+            break;
+        }
+    }
+
+    public void registerMigrationDescriptor(MigrationDescriptor descriptor) {
+        registry.addContribution(descriptor);
+        log.info("Registered migration: " + descriptor.id);
+    }
+
+    public void unregisterMigrationDescriptor(MigrationDescriptor descriptor) {
+        registry.removeContribution(descriptor);
+        log.info("Unregistered migration: " + descriptor.id);
+    }
+
+    @Override
+    public int getApplicationStartedOrder() {
+        return KeyValueServiceImpl.APPLICATION_STARTED_ORDER + 10;
+    }
+
+    protected KeyValueStore getKeyValueStore() {
+        KeyValueService service = Framework.getService(KeyValueService.class);
+        Objects.requireNonNull(service, "Missing KeyValueService");
+        return service.getKeyValueStore(KEYVALUE_STORE_NAME);
+    }
+
+    @Override
+    public MigrationStatus getStatus(String id) {
+        MigrationDescriptor descr = registry.getMigrationDescriptor(id);
+        if (descr == null) {
+            return null; // migration unknown
+        }
+        KeyValueStore kv = getKeyValueStore();
+        String state = kv.getString(id);
+        if (state != null) {
+            return new MigrationStatus(state);
+        }
+        String step = kv.getString(id + STEP);
+        if (step == null) {
+            // migration never started, assumed at first state
+            state = descr.states.keySet().iterator().next(); // LinkedHashMap guarantees order
+            return new MigrationStatus(state);
+        }
+        long startTime = Long.parseLong(kv.getString(id + START_TIME));
+        long pingTime = Long.parseLong(kv.getString(id + PING_TIME));
+        String progressMessage = kv.getString(id + PROGRESS_MESSAGE);
+        long progressNum = Long.parseLong(kv.getString(id + PROGRESS_NUM));
+        long progressTotal = Long.parseLong(kv.getString(id + PROGRESS_TOTAL));
```

Comment:
to be consistent the read need to be done atomicly with the same write lock

Reply:
Yes but I don't want the read path to need locks, as it can be used in polling. The idea is that if the "state" key is present then that's the status, otherwise the "step" and other keys determine the running status, in which case atomicity is not really important. I'll try to rework this to be clearer. The lock is here just to allow atomic transition between "state set and not step" and the opposite.